applications/tari_indexer/src/substate_manager.rs

```diff
@@ -260,35 +263,135 @@ impl SubstateManager {
         Ok(nfts)
     }
 
-    pub async fn get_event_from_db(
-        &self,
-        template_address: TemplateAddress,
-        tx_hash: PayloadId,
-    ) -> Result<Vec<EventData>, anyhow::Error> {
-        let mut tx = self.substate_store.create_read_tx()?;
-        let events = tx.get_events(template_address, tx_hash)?;
-        Ok(events)
-    }
-
     pub async fn save_event_to_db(
         &self,
+        component_address: ComponentAddress,
         template_address: TemplateAddress,
         tx_hash: PayloadId,
         topic: String,
         payload: HashMap<String, String>,
+        version: u64,
     ) -> Result<(), anyhow::Error> {
         let mut tx = self.substate_store.create_write_tx()?;
         let new_event = NewEvent {
+            component_address: Some(component_address.to_string()),
             template_address: template_address.to_string(),
             tx_hash: tx_hash.to_string(),
             topic,
             payload: payload.to_json().expect("Failed to convert to JSON"),
+            version: version as i32,
         };
-        tx.save_events(new_event)?;
+        tx.save_event(new_event)?;
         tx.commit()?;
         Ok(())
     }
 
+    pub async fn scan_events_for_transaction(&self, tx_hash: Hash) -> Result<Vec<Event>, anyhow::Error> {
+        let mut events = vec![];
+        {
+            let mut tx = self.substate_store.create_read_tx()?;
+            let stored_events = match tx.get_events_for_transaction(PayloadId::from_array(tx_hash.into_array())) {
+                Ok(events) => events,
+                Err(e) => {
+                    info!(
+                        target: LOG_TARGET,
+                        "Unable to get events for transaction: {:?} with error {}", tx_hash, e
+                    );
+                    return Err(e.into());
+                },
+            };
+            events.extend(stored_events);
+        }
+
+        let mut events = events
+            .iter()
+            .map(|e| Event::try_from(e.clone()))
+            .collect::<Result<Vec<Event>, anyhow::Error>>()?;
+
+        if events.is_empty() {
+            let network_events = self.substate_scanner.get_events_for_transaction(tx_hash).await?;
+            events.extend(network_events);
+        }
+
+        Ok(events)
```

Comment:
nit: simplified and removes allocations
```suggestion
        let events = {
            let mut tx = self.substate_store.create_read_tx()?;
            tx.get_events_for_transaction(PayloadId::from_array(tx_hash.into_array()))?
        };

        let mut events = events
            .into_iter()
            .map(|e| Event::try_from(e))
            .collect::<Result<Vec<Event>, anyhow::Error>>()?;

        // If we have no events locally, fetch from the network if possible. 
        if events.is_empty() {
            let events = self.substate_scanner.get_events_for_transaction(tx_hash).await?;
            return Ok(events)
        }

        Ok(events)
```

Reply:
Awesome, thanks !